### DISCUSSION TITLE
How to build search functionality for my blog entries?

### DISCUSSION THREAD

**Discussion by syang** [react-page/react-page#1034] [Q&A]
I am using react-page as my blog editors, and we are using DynamoDb to persist the content.

I'd like to have a powerful search functionality (either using ElasticSearch or Algolia.), anyone has the experience on this use case? (There are a lot of tag info in the POST payload -- which made my index very noisy)

**Comment by macrozone** [comment]
good question!

you probably tried to add the whole JSON `Value` into search index, which is noicy.

you could instead extract the text-content of the `Value` by iterating over cells and rows recursivly, check which plugin any cell uses and extract text from each cell.

In many cases its probably enough to extract text from slate (the rich text plugin), There is a small helper function that extracts text from slate's `Node[]`: https://github.com/react-page/react-page/blob/543f3fb50ca35075da1d53087993e15aca0931e4/packages/plugins/content/slate/src/utils/getTextContent.ts, but its not exposed yet.

**Reply by syang** [reply]
@macrozone I did not know react-page is built on top of slate.js, is it? or you guys just partially used some slate.js utilities?

Back to my original question, here is our current architecture design (and please correct me if it is not the right choice):

- we use react-page in our front-end (a react app) to provide article editing / rendering functionality;
- when the user clicks the **Save** button, we make a POST call to /articles (the whole blob, i.e. the whole JSON `Value` is posted back to our backend)

Now, our backend API does the following when it receives the POST call:

1. it saves the whole JSON `Value` into database without any peeking the actual content -- in this way, we can easily serve the GET (which will be serving the article back to the front-end)
2. it make an ElasticSearch call to with the whole JSON `Value` payload, to index the article -- so that the content is searchable

Are you suggesting that I should take the snippet logic you shared above in Step 2 to extract the actual content?

**Comment by macrozone** [comment]
hi @syang,

i now added a function that can extract plaintext from the value:

https://react-page.github.io/docs/#/utils?id=gettextcontents-get-raw-text-contents
